python/lsst/ap/association/diaPipe.py

```diff
@@ -532,50 +693,93 @@ def run(self,
-        # Apdb.
+        Parameters
+        ----------
+        diaObjects : `pandas.DataFrame`
+            Catalog of DiaObjects.
+        updatedDiaObjects : `pandas.DataFrame`
+            Catalog of updated DiaObjects.
+        exposure : `lsst.afw.image.ExposureF`
+            Calibrated exposure differenced with a template to create
+            ``diffIm``.
+        diffIm : `lsst.afw.image.ExposureF`
+            Difference image exposure in which the sources in ``diaSourceCat``
+            were detected.
+        idGenerator : `lsst.meas.base.IdGenerator`
+            Object that generates source IDs and random number generator seeds.
+        preloadedDiaForcedSources : `pandas.DataFrame`
+            Catalog of previously detected forced DiaSources, from the APDB.
+
+        Returns
+        -------
+        diaForcedSources : `pandas.DataFrame`
+            Catalog of calibrated forced photometered fluxes on both the
+            difference and direct images at DiaObject locations.
+        """
+        # Force photometer on the Difference and Calibrated exposures using
+        # the new and updated DiaObject locations.
+        diaForcedSources = self.diaForcedSource.run(
+            diaObjects,
+            updatedDiaObjects.loc[:, "diaObjectId"].to_numpy(),
+            exposure,
+            diffIm,
+            idGenerator=idGenerator)
+        # columns "ra" and "dec" are required for spatial sharding in Cassandra
+        diaForcedSources.rename(columns={"coord_ra": "ra", "coord_dec": "dec"}, inplace=True)
         self.log.info(f"Updating {len(diaForcedSources)} diaForcedSources in the APDB")
         diaForcedSources = convertTableToSdmSchema(self.schema, diaForcedSources,
                                                    tableName="DiaForcedSource",
                                                    )
+        diaForcedSources = self.mergeCatalogs(preloadedDiaForcedSources, diaForcedSources,
+                                              "preloadedDiaForcedSources")
+        if self.testDataFrameIndex(diaForcedSources):
+            self.log.warning(
+                "Duplicate DiaForcedSources created after merge with "
+                "history and new sources. This may cause downstream "
+                "problems. Dropping duplicates.")
+            # Drop duplicates via index and keep the first appearance.
+            # Reset due to the index shape being slight different than
+            # expected.
+            diaForcedSources = diaForcedSources.groupby(
+                diaForcedSources.index).first()
+            diaForcedSources.reset_index(drop=True, inplace=True)
+            diaForcedSources.set_index(
+                ["diaObjectId", "diaForcedSourceId"],
+                drop=False,
+                inplace=True)
+        return diaForcedSources
+
+    @timeMethod
+    def writeToApdb(self, associatedDiaSources, diaForcedSources, updatedDiaObjects):
```

Comment:
Re-order parameters to match the order supplied to `apdb.store`:
```
def writeToApdb(self, updatedDiaObjects, associatedDiaSources, diaForcedSources):
```
Make sure to also update the order of the parameters in `run` where `self.writeToApdb` is called.